docs/server datastorage.md

```diff
@@ -0,0 +1,115 @@
+# Serverside Data Storage
+
+This document covers some of the patterns and tips and tricks used to communicate with data storage. Communication with the data storage is done through the [`Get`](network%20protocol.md#Get) \ [`Retrieved`](network%20protocol.md#Retrieved), [`Set`](network%20protocol.md#Set) \ [`SetReply`](network%20protocol.md#SetReply) and [`SetNotify`](network%20protocol.md#SetNotify) commands described inside the [Network Protocol](network%20protocol.md). The data storage is meant to preserve some data on the server that spans across the lifecycle of the server and is shared with all clients (on any team). 
+
+## Keys
+The data storage works by keys and is internally stored as a dictionary. Note that the whole data storage is accessible by all clients so if you want to store data specifically for your team, game or slot you will have to make your key unique; this is often done by adding the team number, game name or slot id to the key. Some examples:
+
+| EneryLink{team} |                   |
+| ---             | ---               |
+| `EnergyLink1`   | Energy for team 1 |
+| `EnergyLink2`   | Energy for team 2 |
+
+| GiftBoxes;{team} |                                            |
+| ---              | ---                                        |
+| `GiftBoxes;0`    | Giftbox metadata for all giftboxes on team 0 |
+| `GiftBoxes;1`    | Giftbox metadata for all giftboxes on team 1 |
+
+
+| GiftBox;{team};{slot} |                                             |
+| ---                   | ---                                         |
+| `GiftBox;0;1`         | Slot specific giftbox for slot 1 on team 0  |
+| `GiftBox;5;13`        | Slot specific giftbox for slot 13 on team 5 |
+
+## Thread Safety
+Because any client can read and write to the data storage at any time, you can never assume that value you read is still the same 1 second later. So an easy mistake is for example retrieving a value using a `Get` command, then processing the value, altering it, and then writing it back to the store with an `Set`-`replace` command. The `Set`-`replace` command will simply override the value discarding any possible changes by any other clients.
+
+There are a few aspects to the data storage that we can use to our advantage to ensure thread safety:
+* Any additional data given to the `Set` command will be passed along to the `SetReply` response. this can be used to "tag" a certain `SetReply` by for-example adding a specific random value to your `Set` then when you receive the `SetReply` you can check that value to match it back up to your `Set` command. 

```

Comment:
```suggestion
* Any additional data given to the `Set` command will be passed along to the `SetReply` response. This can be used to "tag" a certain `SetReply` by, for example, adding a specific random value to your `Set` then when you receive the `SetReply` you can check that value to match it back up to your `Set` command. 

```